### ISSUE TITLE
Signals are not handled correctly

### ISSUE THREAD

**Issue by sdaschner** [payara/docker-payaraserver-full#25]
Hi there,

when running a hello-world app from the Docker image, the Unix signals are not handled properly.
Running Payara in foreground and <kbd>Ctrl</kbd>+<kbd>C</kbd> doesn't stop the server. Similarly, running `docker stop` takes ten seconds -- the default wait time, before Docker sends a SIGKILL.

See https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/

Tested with `payara/server-full:5-SNAPSHOT` and `payara/server-full:latest` (as of today).

**Comment by mikecroft** [comment]
@OndrejM - is this related to the issue with the launcher that we already know about or something different?

**Comment by OndrejM** [comment]
Hi @mikecroft , I've talked to Sebastian right now and he explained all to me.

The issue is that the launcher is a bash script and bash doesn't propagate Ctrl+C signal to the child JVM process. We need to add a trap for Ctrl+C into the script to forward the signal to the JVM process.

Also, the script should end with error if Payara Server ends with error. This is simple, just execute `set -e` in the beginning of the script.

**Comment by OndrejM** [comment]
@sdaschner pointed me to https://github.com/krallin/tini which is a launcher that is capable of killing the launched process and even all its sub-processes correctly.

**Comment by sdaschner** [comment]
Any updates for this? Actually, this is very much needed to properly run Payara in Docker; otherwise devs have to always construct their own base image.

**Comment by poikilotherm** [comment]
Hey guys,

I just stumbled over this when I tried to stop a container after a deployment exception.
Are you guys willing to merge something if I provide a PR for this?

Cheers!
Oliver

**Comment by MattGill98** [comment]
Hi @poikilotherm,

I do plan to work on this issue. My understanding of this issue is that the init isn't being invoked by default (and so when running docker run with `--init` the problem goes away). Is this correct?

Kind regards,

Matt

**Comment by MattGill98** [comment]
Hi @poikilotherm,

I've made [a commit](https://github.com/payara/docker-payaraserver-full/pull/59/commits/d977d2c3b82f88a61aeaee3a619c0003e111d85f) which I think now addresses this issue. Can you confirm this?

Kind regards,

Matt

**Comment by poikilotherm** [comment]
Hey @MattGill98,

I just tested this. Of course your commit fixes this and Glassfish responds to SIGTERM now.

But this comes at a price, when running without "--init":

``` bash
payara@0ede5deedd66:~$ ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
payara       1  166  3.2 8005624 528124 ?      Ssl  13:37   1:48 /docker-java-home/bin/java -cp /opt/payara/appserver/glassfish/modules/glassfish.jar [...]
payara     196  0.4  0.0  19964  3740 pts/0    Ss   13:38   0:00 /bin/bash
payara     204  0.0  0.0  38388  3084 pts/0    R+   13:38   0:00 ps aux
```

I don't think this is an ideal solution, as I don't think that Java will treat Zombies etc. properly. This will be an issue for usage on Kubernetes too, because in recent versions [PID namespace sharing](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) is still beta and Kubernetes is not able to use `--init`.

Most certainly we should bake in direct support for [tini](https://github.com/krallin/tini) and wrap the current CMD in a new entrypoint script.

**Comment by MattGill98** [comment]
Hi @poikilotherm,

Forgive my naivety, but I'm not quite sure I understand the drawback you're suggesting. I understand that zombie processes won't be reaped correctly, and so I agree with adding direct tini support, certainly in the case of kubernetes. I'm happy to make this addition once my current cleanup PR has been merged.

Kind regards,

Matt

**Comment by poikilotherm** [comment]
Hi @MattGill98 and @Pandrex247 I hope my PR can be considered to be merged and added before the release #60 is made...

Any feedback is very much apprecciated! :smile_cat:

**Comment by OndrejM** [comment]
Thank you, @poikilotherm! We really appreciate as this is an important thing to fix we just couldn't find enough time to do it. Thanks!
